chain/network/src/peer/peer_actor.rs

```diff
@@ -1086,6 +1102,32 @@ impl PeerActor {
                     "Received routed message from {} to {:?}.",
                     self.peer_info,
                     msg.target);
+                let for_me = self.network_state.message_for_me(&msg.target);
+                if for_me {
+                    metrics::record_routed_msg_metrics(&self.clock, &msg);
+                }
+
+                // Drop duplicated messages routed within DROP_DUPLICATED_MESSAGES_PERIOD ms
+                let key = (msg.author.clone(), msg.target.clone(), msg.signature.clone());
+                let now = self.clock.now();
+                if let Some(&t) = self.routed_message_cache.get(&key) {
+                    if now <= t + DROP_DUPLICATED_MESSAGES_PERIOD {
+                        self.network_state.config.event_sink.push(Event::RoutedMessageDropped);
+                        tracing::debug!(target: "network", "Dropping duplicated message from {} to {:?}", msg.author, msg.target);
+                        return;
+                    }
+                }
+                if let RoutedMessageBody::ForwardTx(_) = &msg.body {
+                    // Check whenever we exceeded number of transactions we got since last block.
+                    // If so, drop the transaction.
+                    let r = self.network_state.txns_since_last_block.load(Ordering::Acquire);
+                    if r > MAX_TRANSACTIONS_PER_BLOCK_MESSAGE {
+                        return;
```

Comment:
let's also have a metric for this..

Reply:
also.. this seems to be set only to 1000 (so it is independent of number of shards and number of nodes/validators) --- so this (maybe) might explain why we missed some of the approvals..

Reply:
done

Reply:
I've also added a TODO to analyze why do we need it.